### PR TITLE
Chapter 8: Basic Login

### PR DESCRIPTION
**Pull request by puteriffah**

### CODE REVIEW

**Comment by puteriffah** [comment]
ready

File: Gemfile [hunk 1]
```diff
@@ -8,6 +8,7 @@ gem "rails", "~> 6.0.3", ">= 6.0.3.2"
 gem "rails-i18n"
 gem "mysql2"
 gem "bootstrap-sass", "3.4.1"
+gem 'jquery-rails', '4.3.1'
```
Comment (saobv-2223) [inline Gemfile]: using " please

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class SessionsController < ApplicationController
+  def new
+  end
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]:
```suggestion
  def new; end
```

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  def current_user
+    @current_user ||= User.find_by id: session[:user_id]
+  end
+
+  def logged_in?
+    current_user.present?
+  end
+
+  def log_out
+    session.delete(:user_id)
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove (), please

File: app/javascript/packs/application.js [hunk 4]
```diff
@@ -1,4 +1,6 @@
-require('@rails/ujs').start()
-require('turbolinks').start()
-require('@rails/activestorage').start()
-require('channels')
+require("@rails/ujs").start()
+require("turbolinks").start()
+require("@rails/activestorage").start()
+require("channels")
+require("jquery")
+import "bootstrap"
```
Comment (saobv-2223) [inline app/javascript/packs/application.js]: using ' instead of "

File: app/views/layouts/_header.html.erb [hunk 5]
```diff
@@ -6,7 +6,23 @@
...
+            <ul class="dropdown-menu">
+              <li><%= link_to "Profile", current_user %></li>
+              <li><%= link_to "Settings", "#" %></li>
+              <li class="divider"></li>
+              <li><%= link_to "Log Out", logout_path, method: :delete %></li>
+            </ul>
+          </li>
+        <% else %>
+            <li><%= link_to "Log in", login_path %></li>
+              <% end %>
```
Comment (saobv-2223) [inline app/views/layouts/_header.html.erb]: wrong indenting, fix it please

File: app/views/sessions/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    <%= form_for :session, url: login_path do |f| %>
+    <%= f.label :email %>
+    <%= f.email_field :email, class: "form-control" %>
+    <%= f.label :password %>
+    <%= f.password_field :password, class: "form-control" %>
+    <%= f.submit "Log in", class: "btn btn-primary" %>
+  <% end %>
+  <p><%= t "new_user?" %><%= link_to t("sign_up_now!"), signup_path %></p>
+  </div>
+  </div>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]:
```suggestion
</div>
```

File: app/views/sessions/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  <h1>Log in</h1>
+  <div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for :session, url: login_path do |f| %>
+    <%= f.label :email %>
+    <%= f.email_field :email, class: "form-control" %>
+    <%= f.label :password %>
+    <%= f.password_field :password, class: "form-control" %>
+    <%= f.submit "Log in", class: "btn btn-primary" %>
+  <% end %>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]:
```suggestion
      <%= f.label :email %>
      <%= f.email_field :email, class: "form-control" %>
      <%= f.label :password %>
      <%= f.password_field :password, class: "form-control" %>
      <%= f.submit "Log in", class: "btn btn-primary" %>
    <% end %>
```

File: config/routes.rb [hunk 8]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+
+  get 'sessions/new'
```
Comment (saobv-2223) [inline config/routes.rb]: move it into scope locale and using "

File: e -i HEAD~2 [hunk 9]
```diff
@@ -0,0 +1,11 @@
+[33m442009e[m[33m ([m[1;36mHEAD -> [m[1;32mchapter_8[m[33m)[m first chapt8
+[33md9161dc[m init project
```
Comment (saobv-2223) [inline e -i HEAD~2]: delete this file, please

**Comment by puteriffah** [comment]
ready

File: app/models/user.rb [hunk 10]
```diff
@@ -21,4 +21,5 @@ class User < ApplicationRecord
   def downcase_email
     email.downcase!
   end
+  before_save { email.downcase! }
```
Comment (manhnt-0960) [inline app/models/user.rb]: remove this line because don't use

File: app/views/sessions/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<% provide :title, "log in" %>
+  <h1>Log in</h1>
+  <div class="row">
+  <div class="col-md-6 col-md-offset-3">
+      <%= form_for :session, url: login_path do |f| %>
```
Comment (manhnt-0960) [inline app/views/sessions/new.html.erb]: indent 2 spaces, pls

**Comment by puteriffah** [comment]
ready

File: app/assets/stylesheets/sessions.scss [hunk 12]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/sessions.scss]: remove file that never used

File: app/views/layouts/_header.html.erb [hunk 13]
```diff
@@ -4,9 +4,25 @@
...
+            <ul class="dropdown-menu">
+              <li><%= link_to "Profile", user_path(id: current_user.id) %></li>
+              <li><%= link_to "Settings", "#" %></li>
+              <li class="divider"></li>
+              <li><%= link_to "Log Out", logout_path, method: :delete %></li>
+            </ul>
+          </li>
+          <% else %>
+            <li><%= link_to "Log in", login_path %></li>
+          <% end %>
```
Comment (saobv-2223) [inline app/views/layouts/_header.html.erb]: indenting only 2 spaces, check again

File: app/views/sessions/new.html.erb (hunk 11 above)
Comment (saobv-2223) [reply app/views/sessions/new.html.erb]: fix it, please

**Comment by puteriffah** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,19 @@
+class SessionsController < ApplicationController
+  def new;  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user&.authenticate params[:session][:password]
+      log_in user
+      redirect_to user_path(id: current_user.id)
+    else
+      flash.now[:danger] = "invalid email and password combination. Please try again"
```
Comment (vanvtt-0952) [inline app/controllers/sessions_controller.rb]: use I18n for `"invalid email and password combination. Please try again"`

File: app/views/layouts/_header.html.erb [hunk 15]
```diff
@@ -4,9 +4,25 @@
       <%= link_to "sample app", "#", id: "logo" %>
       <nav>
         <ul class="nav navbar-nav navbar-right">
-          <li><%= link_to "Home", "#" %></li>
-          <li><%= link_to "Help", "#" %></li>
-          <li><%= link_to "Log in", "#" %></li>
+          <li><%= link_to "Home", root_path %></li>
```
Comment (vanvtt-0952) [inline app/views/layouts/_header.html.erb]: use I18n for `"Home"`
and then selfcheck this file, plz

**Comment by vanvtt-0952** [comment]
merged
Please fix finished my 2 comments, then you may be merged.

**Pull request merged**